br/pkg/lightning/backend/local/local.go

```diff
@@ -1269,76 +965,131 @@ func (local *local) readAndSplitIntoRange(ctx context.Context, engine *Engine, r
+		for _, region := range regions {
+			log.FromContext(ctx).Debug("get region",
+				zap.Binary("startKey", startKey),
+				zap.Binary("endKey", endKey),
+				zap.Uint64("id", region.Region.GetId()),
+				zap.Stringer("epoch", region.Region.GetRegionEpoch()),
+				zap.Binary("start", region.Region.GetStartKey()),
+				zap.Binary("end", region.Region.GetEndKey()),
+				zap.Reflect("peers", region.Region.GetPeers()))
+
+			job := &regionJob{
+				keyRange:        intersectRange(region.Region, Range{start: start, end: end}),
+				region:          region,
+				stage:           regionScanned,
+				engine:          engine,
+				regionSplitSize: regionSplitSize,
+				regionSplitKeys: regionSplitKeys,
+				metrics:         local.metrics,
+			}
 
-ScanWriteIngest:
-	for retry := 0; retry < maxRetryTimes; {
-		if retry != 0 {
+			jobWg.Add(1)
 			select {
-			case <-time.After(time.Second):
 			case <-ctx.Done():
 				return ctx.Err()
+			case jobCh <- job:
 			}
 		}
-		startKey := codec.EncodeBytes([]byte{}, pairStart)
-		endKey := codec.EncodeBytes([]byte{}, nextKey(pairEnd))
-		regions, err = split.PaginateScanRegion(ctx, local.splitCli, startKey, endKey, scanRegionLimit)
-		if err != nil || len(regions) == 0 {
-			log.FromContext(ctx).Warn("scan region failed", log.ShortError(err), zap.Int("region_len", len(regions)),
-				logutil.Key("startKey", startKey), logutil.Key("endKey", endKey), zap.Int("retry", retry))
-			retry++
-			continue ScanWriteIngest
-		}
+	}
+	return nil
+}
 
-		for _, region := range regions {
-			log.FromContext(ctx).Debug("get region", zap.Int("retry", retry), zap.Binary("startKey", startKey),
-				zap.Binary("endKey", endKey), zap.Uint64("id", region.Region.GetId()),
-				zap.Stringer("epoch", region.Region.GetRegionEpoch()), zap.Binary("start", region.Region.GetStartKey()),
-				zap.Binary("end", region.Region.GetEndKey()), zap.Reflect("peers", region.Region.GetPeers()))
-
-			w := local.ingestConcurrency.Apply()
-			err = local.writeAndIngestPairs(ctx, engine, region, pairStart, end, regionSplitSize, regionSplitKeys)
-			local.ingestConcurrency.Recycle(w)
+// startWorker creates a worker that reads from the job channel and processes.
+// startWorker will return nil if it's expected to stop, which means the context
+// is canceled or channel is closed. It will return not nil error when it actively
+// stops.
+func (local *local) startWorker(
+	ctx context.Context,
+	jobCh chan *regionJob,
+	jobWg *sync.WaitGroup,
+) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case job, ok := <-jobCh:
+			if !ok {
+				return nil
+			}
+
+			now := time.Now()
+			if now.Before(job.waitUntil) {
+				duration := job.waitUntil.Sub(now)
+				log.FromContext(ctx).Debug("need to wait before processing this job",
+					zap.Duration("wait", duration))
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(duration):
+				}
+			}
+
+			err := local.writeAndIngestPairs(ctx, job)
```

Comment:
Why not move this logic to the job?

Reply:
For now I want the job to provide single actions, so we may extend more action like scanRegion in future and the worker can call single action like a state machine

Reply:
How about abstracting a worker struct and moving all this logic to the worker?

Reply:
> How about abstracting a worker struct and moving all this logic to the worker?

What about this? I think a struct like `jobWorker` is more clear.

Reply:
That require a lof of work, I prefer we finish this PR timely and left it for future work.

we can do it when integrate engine/backend with LOAD DATA/DDL.